xarray/core/computation.py

```diff
@@ -1387,6 +1387,198 @@ def _get_valid_values(da, other):
+    ...         cartesian=(["cartesian"], ["x", "y", "z"]),
+    ...     ),
+    ... )
+    >>> xr.cross(a, b, "cartesian")
+    <xarray.DataArray (time: 2, cartesian: 3)>
+    array([[-3,  6, -3],
+           [ 3, -6,  3]])
+    Coordinates:
+      * time       (time) int64 0 1
+      * cartesian  (cartesian) <U1 'x' 'y' 'z'
+
+    Cross can used by on Datasets by converting to DataArrays and then
+    back to Datasets:
+
+    >>> ds_a = xr.Dataset(dict(x=("dim_0", [1]), y=("dim_0", [2]), z=("dim_0", [3])))
+    >>> ds_b = xr.Dataset(dict(x=("dim_0", [4]), y=("dim_0", [5]), z=("dim_0", [6])))
+    >>> c = xr.cross(
+    ...     ds_a.to_array("cartesian"), ds_b.to_array("cartesian"), dim="cartesian"
+    ... )
+    >>> ds_c = c.to_dataset(dim="cartesian")
+    >>> print(ds_c)
+    <xarray.Dataset>
+    Dimensions:  (dim_0: 1)
+    Dimensions without coordinates: dim_0
+    Data variables:
+        x        (dim_0) int64 -3
+        y        (dim_0) int64 6
+        z        (dim_0) int64 -3
+
+    See Also
+    --------
+    numpy.cross : Corresponding numpy function
+    """
+
+    all_dims: List[Hashable] = []
+    arrays: List[Any] = [a, b]
+    for i, arr in enumerate(arrays):
+        # TODO: Find spatial dim default by looking for unique
+        # (3 or 2)-valued dim?
+        if dim not in arr.dims:
+            raise ValueError(f"Dimension {dim} not in {arr}.")
+
+        if not 1 <= arr.sizes[dim] <= 3:
+            raise ValueError(
+                "Incompatible dimensions for cross product,\n"
+                "dimension with coords must be 1, 2 or 3."
+            )
+
+        all_dims += [d for d in arr.dims if d not in all_dims]
+
+    if arrays[0].sizes[dim] != arrays[1].sizes[dim]:
+        # Arrays have different sizes. Append zeros where the smaller
+        # array is missing a value, zeros will not affect np.cross:
+        i = 1 if arrays[0].sizes[dim] > arrays[1].sizes[dim] else 0
+
+        if all(
+            # The variable check is only used to make mypy happy:
+            getattr(arr, "coords", False) and not isinstance(arr, Variable)
+            for arr in arrays
+        ):
+            # If the arrays have coords we know which indexes to fill
+            # with zeros:
+            arrays[i] = arrays[i].reindex_like(arrays[1 - i], fill_value=0)
+        elif arrays[i].sizes[dim] == 2:
+            # If the array doesn't have coords we can only infer
+            # that it is composite values if the size is 2:
+            arrays[i] = arrays[i].pad({dim: (0, 1)}, constant_values=0)
+            if is_duck_dask_array(arrays[i].data):
+                arrays[i] = arrays[i].chunk({dim: -1})
+        else:
+            # Size is 1, then we do not know if the array is a constant or
+            # composite value:
+            raise ValueError(
+                "Incompatible dimensions for cross product,\n"
+                "dimension without coords must be 2 or 3."
+            )
```

Comment:
Numpy only pads if there was 1 missing value at the end, so I don't think we need to figure out more than just how it's hardcoded now.

How come you left out the dask chunking? I had some case where the arrays weren't chunked correctly and the `apply_ufunc` crashed.

Reply:
I wasn't sure why it was there so I left it out. I think `apply_ufunc` requires core dimensions not to be chunked, regardless of whether or not there are coordinates associated with them so I'd move that directly in front of `apply_ufunc`.

Reply:
Ah right the rechunking was only after the `pad` because `pad` creates weird chunks. 
```python
a = xr.DataArray([1, 2, 3])
a.chunk().pad(dim_0=(0,1))
Out[52]: 
<xarray.DataArray (dim_0: 4)>
dask.array<concatenate, shape=(4,), dtype=float64, chunksize=(3,), chunktype=numpy.ndarray>
Dimensions without coordinates: dim_0
```

If `a` or `b` had weird chunks I considered that user error and let it crash inside `apply_ufunc`.